### PR TITLE
Implement more robust points to tracks algorithm.

### PR DESCRIPTION
**Pull request by litghost**
The previous point algorithm was more complicated than required, and didn't handle all cases.  The new algorithm is more robust, though it may be less optimum.

A follow on will add a function that converts x-lines and y-lines and points and outputs VPR tracks, along with connections.

In particular, the last doctest in the new function caused the previous algorithm to return a disjoint set of lines.

### CODE REVIEW

File: utils/lib/rr_graph/points.py [hunk 1]
```diff
@@ -744,6 +744,285 @@ def straight_closet(line1, line2):
...
+
+    def __repr__(self):
+        return 'Track(dim={},points={})'.format(repr(self.dim), repr(self.points))
+
+def decompose_points_into_tracks(points):
+    """ This function takes a bag of points and returns a set of x lines and
+        y lines that cover all points, and all lines touch each other.
+
+    This is the first step to forming VPR tracks from points.  VPR tracks have
+    limited length, whereas this function returns lines of infinite length.
```
Comment (mithro) [inline utils/lib/rr_graph/points.py]: Do you have something which converts the infinite length lines back to points?

File: utils/lib/rr_graph/points.py [hunk 2]
```diff
@@ -12,10 +12,10 @@
 from collections import defaultdict
 from collections import namedtuple
 
-from ..asserts import assert_eq
-from ..asserts import assert_type
+from lib.asserts import assert_eq
```
Comment (mithro) [inline utils/lib/rr_graph/points.py]: Why did you change these imports?

File: utils/lib/rr_graph/points.py (hunk 1 above)
Comment (litghost) [reply utils/lib/rr_graph/points.py]: I'll add something in another PR, but this function is only for generating the list of rows and columns required to cover the points.

File: utils/lib/rr_graph/points.py (hunk 2 above)
Comment (litghost) [reply utils/lib/rr_graph/points.py]: Invoking doctest explodes without it.

Comment (mithro) [reply utils/lib/rr_graph/points.py]: Go into the symbiflow-arch-defs/utils directory and run `python -m lib.rr_graph.points`?

File: utils/lib/rr_graph/points.py (hunk 1 above)
Comment (mithro) [reply utils/lib/rr_graph/points.py]: BTW Does this cover all the things in this diagram?

image

Comment (litghost) [reply utils/lib/rr_graph/points.py]: It covers the same test cases that the previous algorithm did.  However, it will generate different tracks.  For example, in the global wire case, it might put the span at the end rather than the middle.  Not sure it matters.

Comment (mithro) [reply utils/lib/rr_graph/points.py]: I think you do actually want to minimize total path length? (IE Put the spine in the middle) The previous version probably just got it right by accident though.....

Comment (litghost) [reply utils/lib/rr_graph/points.py]: Why exactly?  Does VPR have channel distance heuristics?

Comment (mithro) [reply utils/lib/rr_graph/points.py]: In some modes, yes.

**Comment by mithro** [comment]
Otherwise, this change looks good to me! Feel free to merge.

File: utils/lib/rr_graph/points.py (hunk 1 above)
Comment (litghost) [reply utils/lib/rr_graph/points.py]: Okay, if we can about minimizing distances, we need something more general like a minimum spanning tree algo.  I believe https://www.sciencedirect.com/science/article/pii/0166218X9400095U is a suitable algo.  I'll file an issue to see if https://www.sciencedirect.com/science/article/pii/0166218X9400095U is suitable and implement it.
